library/common/jni/jni_interface.cc

```diff
@@ -787,6 +788,58 @@ static void* jvm_on_send_window_available(envoy_stream_intel stream_intel, void*
   return result;
 }
 
+// JvmKeyValueStoreContext
+static envoy_data jvm_kv_store_read(envoy_data key, const void* context) {
+  jni_log("[Envoy]", "jvm_kv_store_read");
+  JNIEnv* env = get_env();
+
+  jobject j_context = static_cast<jobject>(const_cast<void*>(context));
+
+  jclass jcls_JvmKeyValueStoreContext = env->GetObjectClass(j_context);
+  jmethodID jmid_read = env->GetMethodID(jcls_JvmKeyValueStoreContext, "read", "([B)[B");
```

Comment:
It's too bad there is no magic for auto-generating these method signature strings, but c'est la vie.

Reply:
In fact, there is! ...but I never figured out how it might work with Bazel. But `javac` itself has support for generating jni headers based on `native` method signatures in class definitions in java.

Reply:
Yeah, this is bound to bite us at some point. `cxx.rs` automatically generates these which helps avoid typos. If we start introducing Rust at this layer it would be worth considering using it for JNI codegen.